### PR TITLE
cli: Use a better amount during transaction simulation

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

There are issues with the current strategy of simulating a test transaction in the `resolve_spend_message`. If the transaction is creating an account, the test transaction will fail because not enough lamports are sent to the destination.

#### Summary of changes

It's a tricky situation in which we can't always be correct, since there's a chicken-and-egg situation with calculating the fee for spend variants of `All` and `RentExempt` if the sender and the fee payer are the same account.

To get the simulation correct in almost all situations, we simulate with the real transfer amount. But if the fee payer is the sender, we simulate with `0`. But we also add a new variant on `SpendAmount` to cover some minimum amount required that must be transferred. Currently, the only situations in which we have an issue are:

* creating a nonce account
* creating a stake account
* transferring SOL

Those first two have a minimum requirement, so use that. The third works fine with a 0 amount, since it's just a SOL transfer.

### CODE REVIEW

File: cli/src/spend_utils.rs [hunk 1]
```diff
@@ -166,7 +168,25 @@ where
...
+            // `0`, since there are few situations in which `SpendAmount` can
+            // be `All` or `RentExempt` *and also* the from account is the fee
+            // payer.
+            let lamports = match amount {
+                SpendAmount::Some(lamports) => lamports,
+                SpendAmount::AllWithMinimum(lamports) => lamports,
+                SpendAmount::All | SpendAmount::RentExempt if from_pubkey == fee_pubkey => 0,
+                SpendAmount::All => from_balance,
+                SpendAmount::RentExempt => from_balance.saturating_sub(from_rent_exempt_minimum),
+            };
```
Comment (jstarry) [inline cli/src/spend_utils.rs]: I feel like this code would be easier to read if the `if from_pubkey == fee_pubkey` if statement wrapped the `match amount`, what do you think? Also I think that `AllWithMinimum` should probably just use `from_balance` when from != fee payer. Does that make sense?

```suggestion
            let lamports = if from_pubkey == fee_pubkey {
                match amount {
                    SpendAmount::Some(lamports) => lamports,
                    SpendAmount::AllWithMinimum(minimum_balance) => minimum_balance,
                    SpendAmount::All | SpendAmount::RentExempt => 0,
                }
            } else {
                match amount {
                    SpendAmount::Some(lamports) => lamports,
                    SpendAmount::AllWithMinimum(_) | SpendAmount::All => from_balance,
                    SpendAmount::RentExempt => {
                        from_balance.saturating_sub(from_rent_exempt_minimum)
                    }
                }
            };
```

File: cli/src/spend_utils.rs [hunk 2]
```diff
@@ -20,6 +20,8 @@ pub enum SpendAmount {
     All,
     Some(u64),
     RentExempt,
+    // will move all, but use a minimum for simulation
+    AllWithMinimum(u64),
```
Comment (jstarry) [inline cli/src/spend_utils.rs]: I feel like this variant name is a little unclear, how about this?
```suggestion
    AllForAccountCreation {
        create_account_min_balance: u64,
    }
```

Comment (joncinque) [reply cli/src/spend_utils.rs]: Yeah, that's much better, thanks!

File: cli/src/spend_utils.rs (hunk 1 above)
Comment (joncinque) [reply cli/src/spend_utils.rs]: This is clearer, thanks for the suggestion

**Review by jstarry — APPROVED** [review]

**Pull request merged**